awx/main/tasks/host_metrics.py

```diff
@@ -3,33 +3,80 @@
+            hard_threshold=getattr(settings, 'CLEANUP_HOST_METRICS_HARD_THRESHOLD', 36),
+        )
+        logger.info("Finished cleanup_host_metrics")
 
 
 @task(queue=get_task_queuename)
 def host_metric_summary_monthly():
     """Run cleanup host metrics summary monthly task each week"""
-    if _is_run_threshold_reached(
-        getattr(settings, 'HOST_METRIC_SUMMARY_TASK_LAST_TS', None), getattr(settings, 'HOST_METRIC_SUMMARY_TASK_INTERVAL', 7) * 86400
-    ):
+    if is_run_threshold_reached(getattr(settings, 'HOST_METRIC_SUMMARY_TASK_LAST_TS', None), getattr(settings, 'HOST_METRIC_SUMMARY_TASK_INTERVAL', 7) * 86400):
         logger.info(f"Executing host_metric_summary_monthly, last ran at {getattr(settings, 'HOST_METRIC_SUMMARY_TASK_LAST_TS', '---')}")
         HostMetricSummaryMonthlyTask().execute()
         logger.info("Finished host_metric_summary_monthly")
 
 
-def _is_run_threshold_reached(setting, threshold_seconds):
-    last_time = DateTimeField().to_internal_value(setting) if setting else DateTimeField().to_internal_value('1970-01-01')
+class HostMetricTask:
+    """
+    This class provides cleanup task for HostMetric model.
+    There are two modes:
+    - soft cleanup (updates columns delete, deleted_counter and last_deleted)
+    - hard cleanup (deletes from the db)
+    """
+
+    def cleanup(self, soft_threshold=None, hard_threshold=None):
+        """
+        Main entrypoint, runs either soft cleanup, hard cleanup or both
 
-    return (now() - last_time).total_seconds() > threshold_seconds
+        :param soft_threshold: (int)
+        :param hard_threshold: (int)
+        """
+        if soft_threshold is not None:
+            self.soft_cleanup(soft_threshold)
+        if hard_threshold is not None:
+            self.hard_cleanup(hard_threshold)
+
+        settings.CLEANUP_HOST_METRICS_LAST_TS = now()
+
+    @staticmethod
+    def soft_cleanup(threshold=None):
+        if not threshold:
+            threshold = getattr(settings, 'CLEANUP_HOST_METRICS_SOFT_THRESHOLD', 12)
+
+        last_automation_before = now() - relativedelta(months=threshold)
+        logger.info(f'cleanup_host_metrics: soft-deleting records last automated before {last_automation_before}')
+        HostMetric.active_objects.filter(last_automation__lt=last_automation_before).update(
+            deleted=True, deleted_counter=F('deleted_counter') + 1, last_deleted=now()
+        )
+
+    @staticmethod
+    def hard_cleanup(threshold=None):
+        if not threshold:
+            threshold = getattr(settings, 'CLEANUP_HOST_METRICS_HARD_THRESHOLD', 36)
+
+        last_deleted_before = now() - relativedelta(months=threshold)
+        logger.info(f'cleanup_host_metrics: hard-deleting records last deleted before {last_deleted_before}')
+        queryset = HostMetric.objects.filter(deleted=True, last_deleted__lt=last_deleted_before)
+        queryset._raw_delete(queryset.db)
```

Comment:
Given what I know about the `HostMetric` model I don't _think_ this should really behave differently from the normal `.delete()` bulk delete, but I don't care a great deal either way.

Reply:
I'm not sure about that, if you're 100% sure that `delete()` won't generate tons of SQL queries or query like `DELETE FROM .. WHERE id IN(...)` etc. which is problematic for 10000+ rows, I can change it. But it's not obvious from the documentation.

Reply:
If `.delete()` did generate a ton of SQL queries like that, then that would be a bit of a red flag that bypassing it could cause an integrity problem (which could be fixed by model changes), but I just have to see for myself using `awx-manage shell_plus --print-sql`

```python
In [20]: HostMetric.objects.filter(deleted=True, last_deleted__lt=hm.last_deleted + timedelta(seconds=1)).delete()
DELETE
  FROM "main_hostmetric"
 WHERE ("main_hostmetric"."deleted" AND "main_hostmetric"."last_deleted" < '2023-08-03 19:42:41.740404+00:00'::timestamptz)

Execution time: 0.001741s [Database: default]
Out[20]: (1, {'main.HostMetric': 1})
```

This gives about the _very basic_ level of validation I needed to say we _don't_ have any red flags, and that either approach should give the same behavior.

To wrap this up, my own _preference_ would be the normal .delete() syntax, but I'm fairly confident behavior will be correct either way and :+1: either way.